src/test/java/nl/ordina/jobcrawler/service/SkillServiceTest.java

```diff
@@ -0,0 +1,298 @@
+    }
+
+
+    @Test
+    void getExistingSkillByName() {
+        /* uses from repository methods: findByName */
+        Mockito.when(skillRepositoryMock.findByName(skillName)).thenReturn(Optional.ofNullable(skill));
+        assertThat(skillService.getSkillByName(skill).get()).isEqualTo(skill);
+    }
+
+    @Test
+    void getNonExistingSkillByName() {
+        /* uses from repository methods: findByName */
+        Mockito.when(skillRepositoryMock.findByName(skillName)).thenReturn(Optional.empty());
+        assertFalse(skillService.getSkillByName(skill).isPresent());
+    }
+
+    @Test
+    void getVacanciesBySkill() {
+        /* uses from repository methods: findByName */
+        Mockito.when(skillRepositoryMock.findByName(skillName)).thenReturn(Optional.ofNullable(skill));
+        Set<Vacancy> result = skillService.getVacanciesBySkill(skillName);
+        assertTrue(result.contains(vacancy)
+                && result.size() == 1);
+    }
+
+    @Test
+    void linkToExistingSkills() {
+        /* uses from repository methods: findByName */
+        Mockito.when(skillRepositoryMock.findByName(skillName)).thenReturn(Optional.ofNullable(skill));
+
+        // make a new skill that only has the name of the existing skill
+        Skill newSkill = Skill.builder()
+                .name(skillName)
+                .vacancies(new HashSet<Vacancy>())
+                .build();
+        Set<Skill> newSkills = new HashSet<Skill>();
+        newSkills.add(newSkill);
+
+        Set<Skill> result = skillService.linkToExistingSkills(newSkills);
+
+        // check if the result contains the skill saved in the "database", and only that
+        assertTrue(result.contains(skill) && result.size() == 1);
+    }
+
+    @Test
+    void deleteSkillsIfNoRelations() {
+        /* uses from repository methods: deleteById, countRelationsById */
+
+        // when the method deleteById gets called with the id of the existing skill, delete that skill from the database
+        Mockito.doAnswer(invocationOnMock -> {
+            skillDB.remove(skill);
+            return null;
+        }).when(skillRepositoryMock).deleteById(skill.getId());
+
+        // when the method countRelationsById gets called with the id of the existing skill,
+        // return 0 indicating that it has no relations
+        Mockito.when(skillRepositoryMock.countRelationsById(skill.getId())).thenReturn(0);
+
+        Set<Skill> skillSet = new HashSet<Skill>();
+        skillSet.add(skill);
+
+        skillService.deleteSkillsIfNoRelations(skillSet);
+
+        assertTrue(skillDB.isEmpty());
+    }
+
+    @Test
+    void addExistingSkillsToVacancy() {
+        /* uses from repository methods: findByName */
+
+        // make new skill
+        Skill newSkill = Skill.builder()
+                .id(UUID.randomUUID())
+                .name("skill2")
+                .vacancies(new HashSet<Vacancy>())
+                .build();
```

Comment:
This pattern appears a couple of times in this test. Consider creating a factory method that takes a skillName, and returns a skill with that name. We might even move that method to Skill itself.

Reply:
added in https://github.com/mostafa-221/jobcrawler-backend/pull/17/commits/c12ef85f064414ae2824339968ebc70111bacf3c